Retrieve Modulemd documents from Koji

OrionStar25 opened the pull request:
Continuing #335 here. 

List of tasks taken care of in this PR:

1. Get Koji tags.
2. Query Koji using these tags and get tagged builds from Koji.
3. Retrieve YAML documents from each build.
4. Create a ModulemdIndex object from each YAML document.

OrionStar25 commented on ModulemdTranslationHelpers/Utils.py:
```diff
@@ -1,14 +1,97 @@
-    from gi.repository import Modulemd
-except ImportError:
-    sys.exit(77)
+gi.require_version('Modulemd', '2.0')
```
> @sgallagher : This whole try/except block should be removed, actually. The sys.exit(77) thing is only intended for tests. If the translation helpers don't have python-libmodulemd, it should just outright fail.

I actually had done that to escape the error generated by `pycodestyle` saying all imports should be on top of the file.

sgallagher replied:
You can fix that with a `# noqa` line, which is appropriate in this case.

sgallagher commented on ModulemdTranslationHelpers/Utils.py:
```diff
@@ -1,14 +1,97 @@
-    from gi.repository import Modulemd
-except ImportError:
-    sys.exit(77)
+gi.require_version('Modulemd', '2.0')
```
```suggestion
gi.require_version('Modulemd', '2.0')  # noqa
```
(Note that it must be over by two spaces from the end of the line or pycodestyle will complain).

sgallagher commented:
I think this looks good, but I'd like to see you write some tests for it. Build the tests around the `f29` tag, because that release is "frozen" and will always have the same content.

OrionStar25 commented:
- [ ] I'll have to look into how to cache the results (I still haven't gotten 1 successful call yet) [TODO]

- [ ] I was wondering if it's possible to have an `.equals()` for `ModulemdIndex` objects as well. Right now I've attempted to check if `ModulemdStream` objects are identical for each module.

sgallagher commented:
> * [ ]  I was wondering if it's possible to have an `.equals()` for `ModulemdIndex` objects as well. Right now I've attempted to check if `ModulemdStream` objects are identical for each module.

Probably the easiest way to do this would be:
```
self.assertEquals(idx.dump_to_string(), index.dump_to_string)
```
That said, this isn't going to work in this case. The YAML in `f29.yaml` contains modulemd-defaults objects, but the YAML in Koji will only contain module stream objects. So those two indices will never be equal.

If you want to check equality like that, you'll need to retrieve that YAML and save it to the test_data as a new filename and compare it that way.

OrionStar25 commented:
@sgallagher Added the case for empty index. Not quite sure about the other case (commented code).

Conan-Kudo commented:
Doesn't this add Koji as a dependency to libmodulemd now?

sgallagher commented:
> Doesn't this add Koji as a dependency to libmodulemd now?

No, this will be a separate python library shipped out of the libmodulemd sources. libmodulemd itself (and python-libmodulemd) will *not* have a koji dependency.

sgallagher commented on ModulemdTranslationHelpers/Utils.py:
```diff
@@ -1,14 +1,111 @@
+class KojiMock:
+
+    def __init__(self, url):
+        self.session = koji.ClientSession(baseurl=url)
```
The whole point of creating a mock object is to *not* actually use `koji.ClientSession`. You instead create an object that simulates it.

sgallagher commented on ModulemdTranslationHelpers/Utils.py:
```diff
@@ -1,14 +1,111 @@
+        self.session = koji.ClientSession(baseurl=url)
+
+    def listTagged(self, tag):
+        return self.session.listTagged(tag)
```
This shouldn't be calling out to koji. You just want to have this return a fixed list of objects in the format that `listTagged()` would return.

sgallagher commented on ModulemdTranslationHelpers/Utils.py:
```diff
@@ -1,14 +1,111 @@
+        return self.session.listTagged(tag)
+
+    def getBuild(self, build_id):
+        return self.session.getBuild(build_id)
```
Same here. What you want here is a `dict()` keyed on `build_id` that has values in the format of `getBuild()`'s return type.

OrionStar25 commented:
@sgallagher 
My index is not being created and I'm not able to understand how to debug this. Is this approach correct though?

sgallagher commented on modulemd/tests/ModulemdTests/translation_helpers.py:
```diff
@@ -137,6 +138,36 @@ def test_translations_from_catalog(self):
+        self.assertIsNotNone(index)
+        mock_session.listTagged.assert_called_once_with('f29')
+        module_names = index.get_module_names()
+        self.assertFalse(module_names)
```
What is this supposed to mean?

OrionStar25 replied:
This test says that a `ModulemdIndex` object will me created but it would be empty.

OrionStar25 replied:
I added a test for a case when `listTagged` wouldn't return anything. In that case, there would be an empty `ModulemdIndex` object.

sgallagher commented on modulemd/tests/ModulemdTests/translation_helpers.py:
```diff
@@ -137,6 +138,36 @@ def test_translations_from_catalog(self):
+        self.assertTrue(len(failures) == 0)
+
+         # Test when koji returns proper modules
+        index = Utils.get_index_from_tags(mock_session, tags)
```
Before calling into `Utils.get_index_from_tags()`, you need to set up the mock functions for `mock_session`. For example:

```python3
        koji_session_mock = Utils.KojiSessionMock()

        mock_session.listTagged.side_effect = koji_session_mock.listTagged
        mock_session.getBuild.side_effect = koji_session_mock.getBuild
```

OrionStar25 commented on ModulemdTranslationHelpers/Utils.py:
```diff
@@ -1,14 +1,136 @@
+class KojiSessionMock:
+
+    def __init__(self):
+        with open('../modulemd/tests/test_data/f29.yaml', 'r') as file:
```
> A proper test here would return just a single module stream document per build_id

I didn't quite understand this bit. Am I not assigning a document only after receiving the `build_id`?

sgallagher replied:
I mean that you shouldn't use f29.yaml for this test, because we actually want the lookups to give us back individual module stream objects (so a single-document YAML) rather than the complete repodata.

OrionStar25 replied:
@sgallagher 
So I have created a new test file which has the sum of 2 individual module stream objects. I hope this is what you meant!

OrionStar25 replied:
I've added 2 module stream objects of `f29` that happen to be the first 2 strings in all the `.po` files.

OrionStar25 commented:
@sgallagher I tried to add `koji` as a dependency in each of the files, and it said no target found. So I removed the import. It then said `No module named requests`. so I added the corresponding module looking at the official package manager for each distro. Still, the module is not able to install only.

OrionStar25 commented:
Okay, the error over here still says `No koji`, which is different from the local error :confused:

OrionStar25 commented:
@sgallagher These are all the different errors:

Mageia: `No module name gi` How was this working before then?
Arch Linux: `target not found: koji`  Its [here](https://aur.archlinux.org/packages/koji/)
OpenSuse: `No provider of 'koji' found.` 
Fedora 30: `cannot import name mock` (for python2) 
Fedora 29: `Failed to synchronize cache for repo 'fedora-modular'`
Fedora Rawhide: `Unable to find a match: python2-koji` Its [here](https://apps.fedoraproject.org/packages/python2-koji)

sgallagher commented:
> ~Okay, the error over here still says `No koji`, which is different from the locally-run docker error~
> I'm getting this same error locally. And opensuse apparently doesn't have any package like `koji/python koji`

Koji is a Fedora tool. It is not expected to exist on other platforms; none of the Translation tests should be run anywhere but Fedora and CentOS. This is part of why I want to move the TranslationHelpers code into its own section in the source hierarchy.

sgallagher commented:
> @sgallagher These are all the different errors:
> 
> Mageia: `No module name gi` How was this working before then?
> Arch Linux: `target not found: koji` Its [here](https://aur.archlinux.org/packages/koji/)
> OpenSuse: `No provider of 'koji' found.`

The above are not expected to have Koji and shouldn't try to run these tests (or build this python module).

> Fedora 30: `cannot import name mock` (for python2)
Python 2 is dead on Fedora 30+. Only Python 3 is supported.

> Fedora 29: `Failed to synchronize cache for repo 'fedora-modular'`
That's a temporary failure; restarting the test should complete fine.

> Fedora Rawhide: `Unable to find a match: python2-koji` Its [here](https://apps.fedoraproject.org/packages/python2-koji)

Rawhide doesn't build python 2 packages anymore except for a very small number of critical ones.

OrionStar25 commented:
@sgallagher Travis is failing and the logs aren't showing the error again. `./make_rpms.sh` worked perfectly in mine.

Conan-Kudo commented:
@sgallagher Koji is packaged in Mageia, we just only ship the Python 3 variant. But there's no Koji instance to target yet...

sgallagher commented:
> @sgallagher Travis is failing and the logs aren't showing the error again. `./make_rpms.sh` worked perfectly in mine.

Edit `make_rpms.sh.in` and drop the `> /dev/null` from the `ninja dist` line. Right now it's suppressing the output.

I need to rework that so it prints the output if the process fails; I'll send a PR.

OrionStar25 commented:
@sgallagher Everything is passing now, except for Rawhide like expected.

OrionStar25 commented:
@sgallagher Error for Fedora 29:

> Error: Failed to synchronize cache for repo 'updates'

sgallagher commented:
> @sgallagher Error for Fedora 29:
> 
> > Error: Failed to synchronize cache for repo 'updates'

That's a transient error. I restarted the test.
